core/src/apps/cardano/helpers/tx_builder.py

```diff
@@ -0,0 +1,194 @@
+from micropython import const
+
+from trezor.crypto import hashlib
+from trezor.messages import CardanoAddressType
+from trezor.messages.CardanoAddressParametersType import CardanoAddressParametersType
+
+from apps.common import cbor
+
+from ..address import derive_address_bytes, get_address_bytes_unsafe
+from ..certificates import cborize_certificate
+from .lazy_collection import DICT, LIST, LazyCollection
+
+if False:
+    from trezor.messages.CardanoSignTx import CardanoSignTx
+    from trezor.messages.CardanoTxInput import CardanoTxInput
+    from trezor.messages.CardanoTxOutput import CardanoTxOutput
+    from trezor.messages.CardanoAssetGroup import CardanoAssetGroup
+    from trezor.messages.CardanoToken import CardanoToken
+    from trezor.messages.CardanoTxCertificate import CardanoTxCertificate
+    from trezor.messages.CardanoTxWithdrawal import CardanoTxWithdrawal
+    from .. import seed
+    from typing import Any, Optional
+
+INPUT = const(0)
+OUTPUT = const(1)
+ASSET_GROUP = const(2)
+TOKEN = const(3)
+CERTIFICATE = const(4)
+WITHDRAWAL = const(5)
+
+
+class TxBuilder:
+    def __init__(self, keychain: seed.Keychain, msg: CardanoSignTx):
+        tx_body_map_item_count = 3 + sum(
+            1
+            for entry in [
+                msg.ttl,
+                msg.certificates_count,
+                msg.withdrawals_count,
+                msg.has_auxiliary_data,
+                msg.validity_interval_start,
+            ]
+            if bool(entry)
+        )
+        self.state = [LazyCollection(DICT, tx_body_map_item_count)]
```

Comment:
IMHO "state" is too generic and it left me at first a bit puzzled, why are we wrapping the lazy dict in an array if it isn't serialized as such. Wouldn't something like "activeItemsStack" be more accurate of a name?